### PR TITLE
Feature/uniswap v2 oracle

### PR DESCRIPTION
**Pull request by rudewalt**

### CODE REVIEW

File: packages/periphery/contracts/oracles/UniswapV2Oracle.sol [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+      price = Math.mulDiv(X96ONE, X96ONE, price);
+    }
+  }
+
+  function getMargincallPrice(address quoteToken, address baseToken) external view returns (uint256 price) {
+    address pair = IUniswapV2Factory(factory).getPair(quoteToken, baseToken);
+
+    price = computePrice(pair, pairOptions[pair].secondsAgoLiquidation);
+    if (quoteToken < baseToken) {
+      price = Math.mulDiv(X96ONE, X96ONE, price);
```
Comment (optifat) [inline packages/periphery/contracts/oracles/UniswapV2Oracle.sol:179]: This is an estimation, which will be different from the precise calculation (using price1Cumulative)

File: packages/periphery/contracts/oracles/UniswapV2Oracle.sol [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    // get the observation for the current period
+    uint8 observationIndex = observationIndexOf(block.timestamp);
+    Observation storage observation = pairObservations[pair][observationIndex];
+
+    // we only want to commit updates once per period (i.e. windowSize / granularity)
+    uint timeElapsed = block.timestamp - observation.timestamp;
+    if (timeElapsed > periodSize) {
+      (uint price0Cumulative, , ) = UniswapV2OracleLibrary.currentCumulativePrices(pair);
+      observation.timestamp = block.timestamp;
+      observation.price0Cumulative = price0Cumulative;
```
Comment (optifat) [inline packages/periphery/contracts/oracles/UniswapV2Oracle.sol:84]: so basically you'll need to implement here something like `observation.priceCumulative = token0IsBaseToken ? price0Cumulative : price1Cumulative`

File: packages/periphery/contracts/oracles/UniswapV2Oracle.sol [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+  // e.g. if the window size is 24 hours, and the granularity is 24, the oracle will return the average price for
+  //   the period:
+  //   [now - [22 hours, 24 hours], now]
+  uint8 public immutable granularity;
+  // this is redundant with granularity and windowSize, but stored for gas savings & informational purposes.
+  uint public immutable periodSize;
+
+  // mapping from pair address to a list of price observations of that pair
+  mapping(address => Observation[]) public pairObservations;
+  address[] public pairs;
```
Comment (optifat) [inline packages/periphery/contracts/oracles/UniswapV2Oracle.sol:54]: I suggest the next way of storing `token0IsBaseToken`: `bool token0IsBaseToken = token0IsBaseToken > 0; address pairAddress = address(uint160(abs(pairs[i])));` and extending `pairs[i]` type up to `int256` or something

File: packages/periphery/contracts/oracles/UniswapV2Oracle.sol [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+  uint public immutable periodSize;
+
+  // pair address converted into number,
+  // when pairKey > 0 means pair.token0 - is base token and pair.token1 - is quote token
+  int256[] public pairKeys;
+
+  // mapping from pairKey to a list of price observations of that pair
+  mapping(int256 => Observation[]) public pairObservations;
+  // mapping from pairKey to options for that pair
+  mapping(int256 => PairOracleOptions) public pairOptions;
```
Comment (Overseven) [inline packages/periphery/contracts/oracles/UniswapV2Oracle.sol:59]: Style guide: https://docs.soliditylang.org/en/v0.8.25/style-guide.html#order-of-layout

Inside each contract, library or interface, use the following order:
1. Type declarations
2. State variables
3. Events
4. Errors
5. Modifiers
6. Functions

**Pull request merged**